### PR TITLE
Fix a bug in C-binding causing iterator to return incorrect result

### LINKED ISSUE #9339: After restart, can't read data from iterator.
The application using C-API can't read data from the iterator after restart. RocksDB version is v27.6.3 and the test procedure is as follows.
1. Set RocksDBto use the prefix extractor and bloom filter.
2. Iterate through RocksDB. If RocksDB is empty, then insert 10 keys (0, 1, 2, ...) into the database.
3. Create an iterator, seek key 0 and iterate until the iterator invalid.

Expected behaviour: restart the test program and the third step should read all keys.

C progam:
```
#include "rocksdb/c.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>

int main() {
    rocksdb_t *db;
    rocksdb_options_t *opts = rocksdb_options_create();
    rocksdb_options_set_create_if_missing(opts, 1);

    // set bloom filter
    rocksdb_block_based_table_options_t *block_opts = rocksdb_block_based_options_create();
    rocksdb_filterpolicy_t *filter = rocksdb_filterpolicy_create_bloom_full(8.0);
    rocksdb_block_based_options_set_filter_policy(block_opts, filter);
    rocksdb_options_set_block_based_table_factory(opts, block_opts);
    // set prefix extractor
    rocksdb_slicetransform_t *prefix = rocksdb_slicetransform_create_fixed_prefix(1);
    rocksdb_options_set_prefix_extractor(opts, prefix);

    char *errptr = NULL;
    db = rocksdb_open(opts, "./data", &errptr);
    assert(errptr == NULL);

    rocksdb_readoptions_t *readopts = rocksdb_readoptions_create();
    rocksdb_iterator_t *iter = rocksdb_create_iterator(db, readopts);
    rocksdb_iter_seek_to_first(iter);
    int cnt = 0;
    while (rocksdb_iter_valid(iter)) {
        cnt++;
        rocksdb_iter_next(iter);
    }
    rocksdb_iter_destroy(iter);
    
    if (cnt == 0) {
        rocksdb_writeoptions_t *writeopts = rocksdb_writeoptions_create();
        for (int i = 0; i < 10; i ++) {
            char key = '0' + i;
            rocksdb_put(db, writeopts, &key, 1, "", 0, &errptr);
            assert(errptr == NULL);
        }
        printf("DB is empty, insert 10 keys\n");
    }

    iter = rocksdb_create_iterator(db, readopts);
    rocksdb_iter_seek(iter, "0", 1);
    cnt = 0;
    while (rocksdb_iter_valid(iter)) {
        cnt++;
        rocksdb_iter_next(iter);
    }
    rocksdb_iter_destroy(iter);
    printf("seek and iterator find %d keys\n", cnt);

    return 0;
}
```

Run the above program twice and the output is as follows. After restarting, the iterator cannot read data after seek.
```
ubuntu@instance-20210714-1042:~/rocksdb$ ./ctest
DB is empty, insert 10 keys
seek and iterator find 10 keys
ubuntu@instance-20210714-1042:~/rocksdb$ ./ctest
seek and iterator find 0 keys
```

This C++ program do same thing and works as expected. So I believe this is C-API bug.
```
#include "rocksdb/db.h"
#include "rocksdb/table.h"
#include "rocksdb/filter_policy.h"
#include "rocksdb/slice_transform.h"

#include <iostream>

using namespace std;
using namespace ROCKSDB_NAMESPACE;

int main() {
  DB* db;
  Options options;
  options.create_if_missing = true;
  // set bloom filter
  BlockBasedTableOptions block_opts;
  auto filter = NewBloomFilterPolicy(8.0, false);
  block_opts.filter_policy.reset(filter);
  options.table_factory.reset(NewBlockBasedTableFactory(block_opts));
  // set prefix extractor
  options.prefix_extractor.reset(NewFixedPrefixTransform(1));

  Status s = DB::Open(options, "./data", &db);
  assert(s.ok());

  // iterate through DB
  int cnt = 0;
  auto iter = db->NewIterator(ReadOptions());
  iter->SeekToFirst();
  while (iter->Valid()) {
    cnt++;
    iter->Next();
  }
  delete iter;

  if (cnt == 0) {
    // create some data in empty DB
    for (int i = 0; i < 10; i++) {
      char key = '0' + i;
      auto s = db->Put(WriteOptions(), string(&key, 1), "");
      assert(s.ok());
    }
    cout << "DB is empty, insert 10 keys" << endl;
  }

  // iterate again
  iter = db->NewIterator(ReadOptions());
  cnt = 0;
  iter->Seek("0");
  while (iter->Valid()) {
    cnt++;
    iter->Next();
  }
  delete iter;
  cout << "seek and iterate find " << cnt << " keys" << endl;

  s = db->Close();
  assert(s.ok());

  return 0;
}
```

### PR DESCRIPTION
**Pull request by riversand963**
Fixes #9339

When writing SST file, the name, computed as `prefix_extractor->GetId()` will be written to the properties block.
When the SST is opened again in the future, `CreateFromString()` will take the name as argument and try
to create a prefix extractor object. Without this fix, the C API will pass a `Wrapper` pointer to the underlying
DB's `prefix_extractor`. `Wrapper::GetId()`, in this case, will be missing the prefix length component, causing a
prefix extractor of length 0 to be silently created and used.

Test:
```
make c_test
./c_test
```

### CODE REVIEW

**Comment by riversand963** [comment]
Since the change is similar to what we do with `Wrapper : rocksdb_filterpolicy_t`, would you be available to take a look? @pdillinger ? Thanks!

File: db/c.cc [hunk 1]
```diff
@@ -4611,18 +4611,18 @@ struct Wrapper : public rocksdb_slicetransform_t {
 };
 
 rocksdb_slicetransform_t* rocksdb_slicetransform_create_fixed_prefix(size_t prefixLen) {
-  Wrapper* wrapper = new Wrapper;
+  SliceTransformWrapper* wrapper = new SliceTransformWrapper;
   wrapper->rep_ = ROCKSDB_NAMESPACE::NewFixedPrefixTransform(prefixLen);
   wrapper->state_ = nullptr;
-  wrapper->destructor_ = &Wrapper::DoNothing;
+  wrapper->destructor_ = &SliceTransformWrapper::DoNothing;
```
Comment (mrambacher) [inline db/c.cc:4617]: Not that it is related to this issue, but shouldn't the destructor delete rep?

File: db/c_test.c [hunk 2]
```diff
@@ -2956,6 +2956,51 @@ int main(int argc, char** argv) {
...
+  {
+    rocksdb_close(db);
+    rocksdb_destroy_db(options, dbname, &err);
+    CheckNoError(err);
+
+    rocksdb_options_set_prefix_extractor(
+        options, rocksdb_slicetransform_create_fixed_prefix(1));
+    rocksdb_filterpolicy_t* filter_policy =
+        rocksdb_filterpolicy_create_bloom_full(8.0);
+    rocksdb_block_based_options_set_filter_policy(table_options, filter_policy);
```
Comment (mrambacher) [inline db/c_test.c:2969]: Are these necessary for the test?  Seems like the slice transform is what is necessary, not the filter policy

File: db/c_test.c [hunk 3]
```diff
@@ -2956,6 +2956,51 @@ int main(int argc, char** argv) {
...
   }
 
+  StartPhase("filter_with_prefix_seek");
+  {
+    rocksdb_close(db);
+    rocksdb_destroy_db(options, dbname, &err);
+    CheckNoError(err);
+
+    rocksdb_options_set_prefix_extractor(
+        options, rocksdb_slicetransform_create_fixed_prefix(1));
```
Comment (mrambacher) [inline db/c_test.c:2966]: Does this prefix extractor need to be deleted to prevent a leak?

File: HISTORY.md [hunk 4]
```diff
@@ -16,6 +16,7 @@
 * Fixed a bug causing two duplicate entries to be appended to a file opened in non-direct mode and tracked by `FaultInjectionTestFS`.
 * Fixed a bug in TableOptions.prepopulate_block_cache to support block-based filters also.
 * Block cache keys no longer use `FSRandomAccessFile::GetUniqueId()` (previously used when available), so a filesystem recycling unique ids can no longer lead to incorrect result or crash (#7405). For files generated by RocksDB >= 6.24, the cache keys are stable across DB::Open and DB directory move / copy / import / export / migration, etc. Although collisions are still theoretically possible, they are (a) impossible in many common cases, (b) not dependent on environmental factors, and (c) much less likely than a CPU miscalculation while executing RocksDB.
+* Fixed a bug in C API by adding the missing GetId() method to `Wrapper`, a subclass of `rocksdb_slicetransform_t` (#9343).
```
Comment (mrambacher) [inline HISTORY.md:19]: Wrapper is internal so you do not need to mention it.  Also, you renamed it to SliceTransformWrapper...

File: db/c_test.c (hunk 2 above)
Comment (riversand963) [reply db/c_test.c:2969]: They are both necessary. We use prefix extractor to extract the prefix and look up in the bloom filter. Before this PR, the bloom filter will return "the key does not exist".

File: HISTORY.md (hunk 4 above)
Comment (riversand963) [reply HISTORY.md:19]: Will update.

File: db/c.cc (hunk 1 above)
Comment (riversand963) [reply db/c.cc:4617]: `rocksdb_slicetransform_t` returned by `rocksdb_slicetransform_create_fixed_prefix()` will be managed by a shared_ptr which will eventually call the destructor of `SliceTransformWrapper`, which calls `delete rep_`.

Comment (riversand963) [reply db/c.cc:4617]: This `destructor_` is not used.

File: db/c_test.c (hunk 3 above)
Comment (riversand963) [reply db/c_test.c:2966]: Please see my response to an earlier comment.

**Comment by riversand963** [comment]
Thanks @mrambacher for the review!

**Review by mrambacher — APPROVED** [review]
LGTM.  Thanks for the fix!

**Comment by riversand963** [comment]
Fb internal linter error is not caused by changes in this PR.

**Comment by pdillinger** [comment]
This does suggest to me that moving to 7.0 sooner, and removing some default implementations, is a good idea to surface bugs associated with failure to override.

Thanks for the fix!

**Comment by riversand963** [comment]
@pdillinger Agree. If the old defaults make it difficult to catch bugs, then we should prepare for 7.0 release soon.

**Pull request closed without merging**